onnxruntime/core/providers/cpu/tensor/expand.cc

```diff
@@ -0,0 +1,179 @@
+    auto input_dim = input_dims_iter > -1 ? input_dims[input_dims_iter] : 1;
+    auto output_dim = output_dims[output_dims_iter];
+
+    input_count *= input_dim;
+    output_count *= output_dim;
+
+    if (0 == input_count || 0 == output_count) {
+      return Status::OK();
+    }
+
+    if (input_dim == 1 && output_dim > 1 || output_dims_iter == 0) {
+      --dim_group_start;
+      input_dim_group[dim_group_start] = input_count;
+      output_dim_group[dim_group_start] = output_count;
+      expand_dim_size[dim_group_start] = output_count / input_count / last_dim_size;
+      last_dim_size *= expand_dim_size[dim_group_start];
+    }
+  }
+
+  auto distribute_count = input_dim_group[dim_group_start] / input_dim_group[max_dims_size - 1];
+  std::vector<int64_t> output_offsets(distribute_count, 0);
+  int64_t copy_len = input_dim_group[max_dims_size - 1];
+  auto copy_byte = copy_len * sizeof(T);
+
+  auto distribute_fn =
+    [&](ptrdiff_t i) {
+      auto input_offset = i * copy_len;
+      int64_t output_offset = 0;
+      for (auto j = dim_group_start + 1, remains = input_offset; j < max_dims_size; ++j) {
+        auto current_count = remains / input_dim_group[j];
+        output_offset += current_count * output_dim_group[j];
+        remains = remains % input_dim_group[j];
+      }  //for
+      memcpy(output_data + output_offset, input_data + input_offset, copy_byte);
+      output_offsets[i] = output_offset;
+    };  //distribute_fn
+
+  auto per_thread_tasks =
+    distribute_count / concurrency::ThreadPool::DegreeOfParallelism(context->GetOperatorThreadPool());
+
+  if (per_thread_tasks > 4) {
+    concurrency::ThreadPool::TrySimpleParallelFor(
+      context->GetOperatorThreadPool(),
+      distribute_count,
+      std::move(distribute_fn));
+  } else {
+    for (int64_t i = 0; i < distribute_count; ++i) {
+      distribute_fn(i);
+    }
+  }  //else
+
+  for (auto i = max_dims_size - 1; i >= dim_group_start; --i) {
+    auto copy_fn =
+    [&](ptrdiff_t j) {
+      auto output_offset = output_offsets[j];
+      if (output_offset % output_dim_group[i] == 0) {
+        auto copy_len = output_dim_group[i] / expand_dim_size[i];
+        auto copy_byte = copy_len * sizeof(T);
+        auto output_from = output_data + output_offset;
+        auto output_at = output_from + copy_len;
+        auto output_end = output_from + output_dim_group[i];
+        while (output_at + copy_len <= output_end) {
+          memcpy(output_at, output_from, copy_byte);
+          output_at += copy_len;
+          copy_len <<= 1;
+          copy_byte <<= 1;
+        }  //while
+        while (output_at < output_end) {
+          if (output_at + copy_len <= output_end) {
+            memcpy(output_at, output_from, copy_byte);
+            output_at += copy_len;
+          } else {
+            copy_len >>= 1;
+            copy_byte >>= 1;
+          }
+        }  //while
+      }  //if
+    };  //copy_fn
+    if (per_thread_tasks > 20) {
```

Comment:
A small note indicating how you arrived at this threshold would help future maintainers.

Reply:
tensor of size (128, 1, 128) expand to (128, copy_count, 128):

copy_count | time(us)
---- | ----
8 | 
8 | 
8 |